### PR TITLE
Filedecoratorplugin

### PR DESCRIPTION
**Pull request by bunsenstraat**
Screenshot 2022-07-21 at 15 42 00

This adds 'decorators' to the File Explorer and Tabs.
There are now 3 predefined decorators:
ERROR
WARNING
CUSTOM

the text displayed in the error and warnings and mainly meant for displaying the amount
the comments appear on hover
fileStateIcon is used for custom only, it can be any type of element ( react too ) or a text
the fileStateLabelClass is added to the name of the file
the fileStateIconClass is added to the fileStateIcon
bubble: means the parent folders will be decorated with the fileStateLabelClass

methods are 
clearFileDecorators(path?: string)
setFileDecorators: array of fileDecorators
clearAllFileDecorators

this script demonstrates the functions

```
let decorator: any = {
                path: 'contracts/2_Owner.sol',
                isDirectory: false,
                fileStateType: 'ERROR',
                fileStateLabelClass: 'text-danger',
                fileStateIconClass: '',
                fileStateIcon: '',
                text: '2',
                bubble: true,
                comment: 'error on owner',
            }
let decorator2: any = {
                path: 'contracts/2_Owner.sol',
                isDirectory: false,
                fileStateType: 'CUSTOM',
                fileStateLabelClass: 'text-success',
                fileStateIconClass: 'text-success',
                fileStateIcon: 'U',
                text: '',
                bubble: true,
                comment: 'modified',
            }
remix.call('fileDecorator' as any, 'setFileDecorators', [decorator, decorator2])

decorator = {
                path: 'contracts/1_Storage.sol',
                isDirectory: false,
                fileStateType: 'WARNING',
                fileStateLabelClass: 'text-warning',
                fileStateIconClass: '',
                fileStateIcon: '',
                text: '2',
                bubble: true,
                comment: 'warning on storage',
            }
remix.call('fileDecorator' as any, 'setFileDecorators', decorator)

decorator = {
                path: 'contracts/3_Ballot.sol',
                isDirectory: false,
                fileStateType: 'CUSTOM',
                fileStateLabelClass: '',
                fileStateIconClass: '',
                fileStateIcon: 'customtext',
                text: 'with text',
                bubble: true,
                comment: 'custom comment',
            }
remix.call('fileDecorator' as any, 'setFileDecorators', decorator)
```

### CODE REVIEW

File: apps/remix-ide/src/app/plugins/file-decorator.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+        if (!deepequal(newState, this._fileStates)) {
+            this._fileStates = newState
+            this.emit('fileDecoratorsChanged', this._fileStates)
+        }
+    }
+
+    async clearFileDecorators(owner? : string) {
+        if(!owner) {
+            this._fileStates = []
```
Comment (yann300) [inline apps/remix-ide/src/app/plugins/file-decorator.ts:50]: if `owner` is not set that might indicate a wrong call and it should maybe throw or just do nothing

**Comment by yann300** [comment]
```
let decorator2: any = {
                path: 'contracts/2_Owner.sol',
                isDirectory: false,
                fileStateType: 'CUSTOM',
                fileStateLabelClass: 'text-success',
                fileStateIconClass: 'text-success',
                fileStateIcon: 'U',
                text: '',
                owner: 'code-parser',
                bubble: true,
                comment: 'modified',
            }
```
ideally could be that the current value of owner is set by `this.current` (this is injected by remix-plugin) instead of hardcoding `owner`. That way plugins can't override others.
But if the plugin is meant to be used only internally, I guess that's ok.

File: apps/remix-ide/src/app/plugins/file-decorator.ts (hunk 1 above)
Comment (bunsenstraat) [reply apps/remix-ide/src/app/plugins/file-decorator.ts:50]: no if there's no owner all decorators get cleared, ie when you change workspace.

**Comment by bunsenstraat** [comment]
> ```
> let decorator2: any = {
>                 path: 'contracts/2_Owner.sol',
>                 isDirectory: false,
>                 fileStateType: 'CUSTOM',
>                 fileStateLabelClass: 'text-success',
>                 fileStateIconClass: 'text-success',
>                 fileStateIcon: 'U',
>                 text: '',
>                 owner: 'code-parser',
>                 bubble: true,
>                 comment: 'modified',
>             }
> ```
> 
> ideally could be that the current value of owner is set by `this.current` (this is injected by remix-plugin) instead of hardcoding `owner`. That way plugins can't override others. But if the plugin is meant to be used only internally, I guess that's ok.

yes good idea

**Review by yann300 — APPROVED** [review]

**Review by yann300 — APPROVED** [review]

**Pull request merged**